Phase2

Doinededoidoine opened the pull request.

Doinededoidoine committed:
Phase 2 : Parent_table, get_parent, get_root

Doinededoidoine committed:
Algorithme de Kruskal (ne fonctionne pas)

Bugs :
- Les méthodes nodes() et edges() renvoient une erreur UndefVarError
- root() : problème avec le type de node1 et node2

mtanneau commented on projet/phase2/main2.jl:
```diff
@@ -0,0 +1,44 @@
+import Base.show
+
+here = "C://Users//arthg//github//mth6412b-starter-code//projet//"
+include(joinpath(here, "phase1//graph.jl"))
```
Utilisez plutot la syntaxe suivante (au lieu des lignes 4-5):
```julia
include(joinpath(@__DIR__, "..", "phase1", "graph.jl"))
```
(la macro `@__DIR__` renvoie le dossier dans lequel le fichier courant se trouve)

mtanneau commented on projet/phase1/graph.jl:
```diff
@@ -60,3 +63,18 @@ function show(graph::Graph)
+"""Renvoie les noeuds situés aux extrémités d'une arête d'un graphe donné."""
+function get_sd_nodes(nodes::Vector{Node{T}} where T, edge::AbstractEdge)
+  node1 = nodes[1]
+  node2 = nodes[1]
```
```node2 = nodes[1]``` devrait être ```node2 = nodes[2]```

mtanneau commented:
Il faut corriger une partie de votre code de la phase 1, notamment concernant la représentation des arêtes.

```
mutable struct Edge{T} <: AbstractEdge{T}
  name::String
  s_node::T
  d_node::T
  weight::T
end
```
nous ne pouvez pas forcer `s_node` et `d_node` à avoir le même type que `weight`.
`s_node` et `d_node` sont des nœuds, ils devraient donc être de type `Node{T}` (ou plus generalement `AbstractNode{T}`.
Le poids `weight` est un scalaire, vous pouvez utiliser par exemple un `Int64` (si vous avez la garantie que tous les poids seront toujours entiers) ou `Float64`.

Cela vous permettra de vous dispenser de la fonction `get_sd_nodes` (qui vous fait potentiellement parcourir la liste de tous les nœuds du graphe pour obtenir les extrémités d'une arête !).

Doinededoidoine committed:
- Edge : starting and destination nodes now have type Node{T}
- Fixed previous bugs
- Problem : excessive number of edges in min_tree.

mtanneau commented:
Vous devrez importer votre code de `main1` et `main2` dans `runtests.jl` :
```
include("main2.jl")
include(joinpath(@__DIR__, "..", "phase1", "main1.jl"))
```

mtanneau commented:
Avec ça (plus l’utilisation d'un chemin relatif pour le fichier .tsp), tout tourne à part le dernier test unitaire :
```julia
Test Failed at /mnt/c/Users/mathi/Desktop/mth6412/code/Doinededoidoine/projet/phase2/runtests.jl:54
  Expression: length(edges(min_tree3)) == length(nodes(graph3)) - 1
   Evaluated: 103 == 28
```

Doinededoidoine committed:
runtests : tests unitaires

mtanneau commented:
MAJ: Le code tourne mais j'obtiens (toujours) l'erreur suivante lors des tests :
```julia
Test Failed at /mnt/c/Users/mathi/Desktop/mth6412/code/Doinededoidoine/projet/phase2/runtests.jl:131
  Expression: length(edges(min_tree_4n)) == length(nodes(min_tree_4n)) - 1
   Evaluated: 6 == 3
ERROR: LoadError: There was an error during testing
```

Doinededoidoine committed:
Wrote unitary tests.
Fixed previous bug.
Algorithm still not working : fusion of two distinct connex components 
does not work properly.

mtanneau commented on projet/phase1/edge.jl:
```diff
@@ -14,8 +16,8 @@ Exemple:
-  d_node::T
+  s_node::AbstractNode{T}
+  d_node::AbstractNode{T}
   weight::T
```
`weight` ne doit pas être du type `T` : si `T = Vector{Float64}`, le poids d'une arête serait alors un tableau !

mtanneau commented on projet/phase1/graph.jl:
```diff
@@ -21,13 +24,13 @@ mutable struct Graph{T} <: AbstractGraph{T}
 
 """Ajoute un noeud au graphe."""
-function add_node!(graph::Graph{T}, node::Node{T}) where T
+function add_node!(graph::Graph{T}, node::AbstractNode{T}) where T
```
Gardez `node::Node{T}` : par définition, `graph.nodes` est un tableau de `Node{T}`s, vous ne pouvez pas ajouter un élément de type différent que `Node{T}`.

mtanneau commented on projet/phase1/graph.jl:
```diff
@@ -21,13 +24,13 @@ mutable struct Graph{T} <: AbstractGraph{T}
 
 """Ajoute un arc au graphe."""
-function add_edge!(graph::Graph{T}, edge::Edge{T}) where T
+function add_edge!(graph::Graph{T}, edge::AbstractEdge{T}) where T
```
idem

mtanneau commented on projet/phase1/main1.jl:
```diff
@@ -0,0 +1,28 @@
+include("edge.jl")
+include("graph.jl")
+include("read_stsp.jl")
+using Plots
```
Pas besoin de `using Plots` ici

mtanneau commented on projet/phase1/main1.jl:
```diff
@@ -0,0 +1,28 @@
+  edges_empty = Vector{Edge}()
+  graph = Graph{Int64}("graph", nodes_empty, edges_empty)
+  for (i, node) in enumerate(graph_nodes)
+    new_node = Node{Int64}(string(collect(keys(graph_nodes))[i]), collect(keys(graph_nodes))[i])
```
Vous supprimez l'info contenue dans les nœuds en faisant ça !

mtanneau commented on projet/phase1/read_stsp.jl:
```diff
@@ -1,5 +1,9 @@
 
+include("node.jl")
+include("edge.jl")
+include("graph.jl")
```
Vous n'avez pas besoin des trois `include` ici.

mtanneau commented on projet/phase2/main2.jl:
```diff
@@ -0,0 +1,56 @@
+
+import Base.show
```
Pas besoin d'importer `Base.show`

mtanneau commented on projet/phase2/main2.jl:
```diff
@@ -0,0 +1,56 @@
+"""Renvoie un arbre de recouvrement minimal du graphe symétrique en entrée en
+utilisant l'algorithme de Kruskal. La méthode renvoie un objet de type Graph.
+"""
+function main2(graph::AbstractGraph)
```
Ce code ne correspond pas à ce qu'il y a dans votre rapport

mtanneau commented on projet/phase2/main2.jl:
```diff
@@ -0,0 +1,56 @@
+    edges_copy = copy(edges(graph))
+    # On trie les arêtes par poids croissant.
+    edges_order = sortperm(weight.(edges_copy))
+    edges_copy = edges_copy[edges_order]
```
Vous pouvez directement faire
```julia
sort!(edge_copy, by=weight)
```
ce qui trie le tableau `edge_copy` en place, par ordre croissant de `weight` (ici, `weight` est une fonction)

mtanneau commented on projet/phase2/main2.jl:
```diff
@@ -0,0 +1,56 @@
+    # On trie les arêtes par poids croissant.
+    edges_order = sortperm(weight.(edges_copy))
+    edges_copy = edges_copy[edges_order]
+    min_tree = Graph{Int64}("min_tree", [], [])
```
Si `graph` est un `Graph{T}`, il faut renvoyer un `Graph{T}`, pas un `Graph{Int}`

mtanneau commented on projet/phase2/main2.jl:
```diff
@@ -0,0 +1,56 @@
+        # l'arête à l'arbre de recouvrement minimum.
+        if root(parent_table, node1) != root(parent_table, node2)
+            add_edge!(min_tree, edge)
+            # On veut ensuite savoir quel(s) noeud(s) ajouter à l'arbre.
```
L'arbre couvrant contient exactement _tous_ les nœuds du graphe de départ : vous pouvez les ajouter avant même de commencer l’algorithme, ce qui vous évite tout ce qui suit.

mtanneau commented on projet/phase2/main2.jl:
```diff
@@ -0,0 +1,56 @@
+                child_node = node2
+            end
+            # On met finalement à jour la table des parents.
+            set_parent!(parent_table, child_node, parent_node)
```
L'union des composantes connexes se fait via la racine (cf labo).

mtanneau commented on projet/phase2/parent_table.jl:
```diff
@@ -0,0 +1,62 @@
+    parents::Vector{Node{T}}
+end
+
+ParentTable{T}() where T = ParentTable{T[]}
```
`ParentTable{T}()` est un constructeur (i.e., une fonction), alors que `ParentTable{T[]}` est un type.

mtanneau commented on projet/phase2/parent_table.jl:
```diff
@@ -0,0 +1,62 @@
+            return parent
+        end
+    end
+    parent
```
Il faudrait plutôt soulever une erreur si `node` ne figure pas dans le tableau des enfants.

mtanneau commented on projet/phase2/parent_table.jl:
```diff
@@ -0,0 +1,62 @@
+            return parent_table
+        end
+    end
+    parent_table
```
idem : soulevez une erreur si `node` ne figure pas dans le tableau des enfants.

mtanneau commented on projet/phase2/parent_table.jl:
```diff
@@ -0,0 +1,62 @@
+"""
+function init_parent_table(graph::Graph)
+    graph_nodes_copy = copy(nodes(graph))
+    parent_table = ParentTable{Int64}(nodes(graph), graph_nodes_copy)
```
Cette ligne renverra une erreur dès que `graph` est un objet de type `Graph{T}` avec `T != Int64`.

mtanneau commented on projet/phase2/runtests.jl:
```diff
@@ -0,0 +1,166 @@
+include(joinpath(@__DIR__, "..", "phase1", "main1.jl"))
+include(joinpath(@__DIR__, "main2.jl"))
+
+using Test
```
dupplicate

mtanneau commented on projet/phase2/runtests.jl:
```diff
@@ -0,0 +1,166 @@
+
+using Test
+
+"""Construction des graphes"""
```
N'utilisez des docstrings que pour décrire le comportement d'une fonction _individuelle_.

Ici, utilisez un commentaire normal

mtanneau commented on projet/phase2/runtests.jl:
```diff
@@ -0,0 +1,166 @@
+
+"""Construction des graphes"""
+
+# "Graphe à 3 noeuds"
```
Pas besoin des guillemets à l’intérieur d'un commentaire

mtanneau commented on projet/phase2/runtests.jl:
```diff
@@ -0,0 +1,166 @@
+parent_table_Nn3 = init_parent_table(graph_Nn3)
+
+
+"""Fonctions"""
```
Idem : utilisez un commentaire régulier.

mtanneau reviewed (changes requested):
A corriger : 
* `Edge`
* Algorithme de Kruskal

Vous pouvez grandement améliorer la complexité de votre code.

Doinededoidoine committed:
Correction des phases 1 & 2

Doinededoidoine committed:
Phase 2 fonctionnelle

Doinededoidoine committed:
Implémentation des deux heuristiques d'accélération
